eclair-core/src/main/scala/fr/acinq/eclair/payment/Relayer.scala

```diff
@@ -215,34 +213,34 @@ object Relayer {
   // @formatter:off
   sealed trait NextPayload
   case class FinalPayload(add: UpdateAddHtlc, payload: PerHopPayload) extends NextPayload
-  case class RelayPayload(add: UpdateAddHtlc, payload: PerHopPayload, nextPacket: Sphinx.Packet) extends NextPayload {
+  case class RelayPayload(add: UpdateAddHtlc, payload: PerHopPayload, nextPacket: OnionRoutingPacket) extends NextPayload {
     val relayFeeMsat: Long = add.amountMsat - payload.amtToForward
     val expiryDelta: Long = add.cltvExpiry - payload.outgoingCltvValue
   }
   // @formatter:on
 
   /**
-    * Parse and decode the onion of a received htlc, and find out if the payment is to be relayed,
+    * Decrypt the onion of a received htlc, and find out if the payment is to be relayed,
     * or if our node is the last one in the route
     *
     * @param add        incoming htlc
     * @param privateKey this node's private key
-    * @return the payload for the next hop
+    * @return the payload for the next hop or an error.
     */
-  def tryParsePacket(add: UpdateAddHtlc, privateKey: PrivateKey): Try[NextPayload] =
-    Sphinx
-      .parsePacket(privateKey, add.paymentHash, add.onionRoutingPacket)
-      .flatMap {
-        case Sphinx.ParsedPacket(payload, nextPacket, _) =>
-          LightningMessageCodecs.perHopPayloadCodec.decode(BitVector(payload)) match {
-            case Attempt.Successful(DecodeResult(perHopPayload, _)) if nextPacket.isLastPacket =>
-              Success(FinalPayload(add, perHopPayload))
-            case Attempt.Successful(DecodeResult(perHopPayload, _)) =>
-              Success(RelayPayload(add, perHopPayload, nextPacket))
-            case Attempt.Failure(cause) =>
-              Failure(new RuntimeException(cause.messageWithContext))
-          }
-      }
+  def decryptPacket(add: UpdateAddHtlc, privateKey: PrivateKey): Either[BadOnion, NextPayload] =
+    Sphinx.PaymentPacket.peel(privateKey, add.paymentHash, add.onionRoutingPacket) match {
+      case Right(p@Sphinx.DecryptedPacket(payload, nextPacket, _)) =>
+        OnionCodecs.perHopPayloadCodec.decode(payload.bits) match {
+          case Attempt.Successful(DecodeResult(perHopPayload, _)) if p.isLastPacket =>
```

Comment:
Should we check that `remainder` is empty?

Reply:
I don't think it's this layer's responsibility.
For 65-bytes frames, the sphinx layer ensures that what we receive in the `Relayer` has the proper format so we will always have an empty remainder.
For variable-length frames, the sphinx layer also guarantees that we'll receive exactly the bytes of the tlv stream, so it should always be empty too.
Do you want to add a `require` to catch bugs (since this is a case that should never happen if the other layers behave properly)?

Reply:
I think it's always re-assuring to add such a `require` check to mention that the other layers should make sure we end up with an empty remainder here.

Reply:
A `require` here would result in an infinite loop. Doing this:

```scala
 OnionCodecs.perHopPayloadCodec.decode(payload.bits) match {
          case Attempt.Successful(DecodeResult(perHopPayload, remainder)) if p.isLastPacket && remainder.isEmpty =>
            Right(FinalPayload(add, perHopPayload))
          case Attempt.Successful(DecodeResult(perHopPayload, remainder)) if remainder.isEmpty =>
            Right(RelayPayload(add, perHopPayload, nextPacket))
          case _ =>
            // Onion is correctly encrypted but the content of the per-hop payload couldn't be parsed.
            Left(InvalidOnionPayload(Sphinx.PaymentPacket.hash(add.onionRoutingPacket)))
        }
```

would correctly return an `InvalidOnionPayload` but it doesn't look very good.

Alternatively:
```scala
 OnionCodecs.perHopPayloadCodec.decode(payload.bits) match {
          case Attempt.Successful(DecodeResult(_, remainder)) if remainder.nonEmpty =>
            Left(InvalidOnionPayload(Sphinx.PaymentPacket.hash(add.onionRoutingPacket)))
          case Attempt.Successful(DecodeResult(perHopPayload, _)) if p.isLastPacket =>
            Right(FinalPayload(add, perHopPayload))
          case Attempt.Successful(DecodeResult(perHopPayload, _)) =>
            Right(RelayPayload(add, perHopPayload, nextPacket))
          case Attempt.Failure(_) =>
            // Onion is correctly encrypted but the content of the per-hop payload couldn't be parsed.
            Left(InvalidOnionPayload(Sphinx.PaymentPacket.hash(add.onionRoutingPacket)))
        }
```
doesn't look much better.

Reply:
What about this?

```scala
OnionCodecs.perHopPayloadCodec.decode(payload.bits) match {
          case Attempt.Successful(DecodeResult(perHopPayload, remainder)) =>
            require(remainder.isEmpty, "there shouldn't be any remaining bytes after decoding sphinx payload")
            if (p.isLastPacket)
              Right(FinalPayload(add, perHopPayload))
            else
              Right(RelayPayload(add, perHopPayload, nextPacket))
          case Attempt.Failure(_) =>
            // Onion is correctly encrypted but the content of the per-hop payload couldn't be parsed.
            Left(InvalidOnionPayload(Sphinx.PaymentPacket.hash(add.onionRoutingPacket)))
        }
```